LoopKit/TemporaryScheduleOverrideHistory.swift

```
@@ -0,0 +1,239 @@
+                        return (index1, index2)
+                    } else {
+                        return nil
+                    }
+                }
+
+        let invalidOverrideIndices = overlappingOverridePairIndices.flatMap { [$0, $1] }
+        guard invalidOverrideIndices.isEmpty else {
+            // Save the invalid event history for debugging.
+            taintedEventLog = recentEvents
+
+            // Wipe only conflicting overrides to retain as much history as possible.
+            recentEvents.removeAll(at: invalidOverrideIndices)
+
+            // Crash deliberately to notify something has gone wrong.
+            preconditionFailure("No overrides should overlap.")
+        }
+    }
+
+    func wipeHistory() {
+        recentEvents.removeAll()
+    }
+}
+
+
+extension OverrideEvent: RawRepresentable {
+    typealias RawValue = [String: Any]
+
+    init?(rawValue: RawValue) {
+        guard
+            let overrideRawValue = rawValue["override"] as? TemporaryScheduleOverride.RawValue,
+            let override = TemporaryScheduleOverride(rawValue: overrideRawValue)
+        else {
+            return nil
+        }
+
+        self.override = override
+
+        if let endDate = rawValue["endDate"] as? Date {
+            self.end = .early(endDate)
+        }
+    }
+
+    var rawValue: RawValue {
+        var raw: RawValue = [
+            "override": override.rawValue
+        ]
+
+        if case .early(let endDate) = end {
+            raw["endDate"] = endDate
+        }
+
+        return raw
+    }
+}
+
+
+extension TemporaryScheduleOverrideHistory: RawRepresentable {
+    public typealias RawValue = [String: [[String: Any]]]
+
+    public convenience init?(rawValue: RawValue) {
+        self.init()
+        if let recentEventsRawValue = rawValue["recentEvents"] {
+            self.recentEvents = recentEventsRawValue.compactMap(OverrideEvent.init(rawValue:))
```

Again, `compactMap` here and 3 lines below technically hides failures.